File: analyze_btsnoop.dart
import 'dart:io';
import 'dart:typed_data';

void main(List<String> args) async {
  if (args.isEmpty) {
    print("Usage: dart analyze_btsnoop.dart <file1> [file2...]");
    return;
  }

  for (var path in args) {
    if (await File(path).exists()) {
      await parseBtsnoop(path);
    } else {
      print("File not found: $path");
    }
  }
}

Future<void> parseBtsnoop(String path) async {
  print("--- Analyzing $path ---");
  final file = File(path);
  final bytes = await file.readAsBytes();
  final ByteData data = bytes.buffer.asByteData();
  int offset = 0;

  if (data.lengthInBytes < 16) {
    print("File too short");
    return;
  }

  // Header: 8 bytes magic ("btsnoop\0"), 4 ver, 4 datalink
  // Check Magic
  // btsnoop\0 = 62 74 73 6E 6F 6F 70 00
  if (bytes[0] != 0x62 || bytes[7] != 0x00) {
    print("Invalid Magic");
    return;
  }

  offset += 8;
  int version = data.getUint32(offset, Endian.big);
  offset += 4;
  int datalink = data.getUint32(offset, Endian.big);
  offset += 4;

  print("Version: $version, DataLink: $datalink (1002=HCI UART)");

  int packetCount = 0;

  while (offset < data.lengthInBytes) {
    if (offset + 24 > data.lengthInBytes) break;

    // Record Header
    int origLen = data.getUint32(offset, Endian.big);
    offset += 4;
    int incLen = data.getUint32(offset, Endian.big);
    offset += 4;
    int flags = data.getUint32(
        offset, Endian.big); // Bit 0: 0=Sent(Com/Evt), 1=Recv?? NO.
    // Packet Flags:
    // Bit 0: Direction (0=Sent, 1=Recv)
    // Bit 1: Command/Data (0=Data, 1=Cmd/Evt)
    offset += 4;
    int drops = data.getUint32(offset, Endian.big);
    offset += 4;
    // Time
    offset += 8;

    // Packet Data
    if (offset + incLen > data.lengthInBytes) break;

    Uint8List packetData = bytes.sublist(offset, offset + incLen);
    offset += incLen;
    packetCount++;

    if (packetData.isEmpty) continue;

    // Parsing HCI UART (H4)
    // Type (1 byte)
    // 01=Cmd, 02=ACL, 04=Event
    int type = packetData[0];

    if (type == 0x02) {
      // ACL
      // ACL Header (4 bytes): Handle(12) | PB(2) | BC(2), TotalLen(2)
      if (packetData.length < 5) continue;

      int handleFlags = packetData[1] | (packetData[2] << 8);
      int connHandle = handleFlags & 0x0FFF;

      // L2CAP Header (4 bytes): Len(2), CID(2)
      // Offset = 1(Type) + 4(ACL) = 5
      int l2capOffset = 5;
      if (packetData.length < l2capOffset + 4) continue;

      int l2capLen =
          packetData[l2capOffset] | (packetData[l2capOffset + 1] << 8);
      int l2capCid =
          packetData[l2capOffset + 2] | (packetData[l2capOffset + 3] << 8);

      if (l2capCid == 0x04) {
        // ATT
        int attOffset = l2capOffset + 4;
        if (attOffset >= packetData.length) continue;

        Uint8List attData = packetData.sublist(attOffset);
        if (attData.isEmpty) continue;

        int opcode = attData[0];
        bool relevant = false;
        String name = "";

        // 0x12 Write Req, 0x52 Write Cmd, 0x1B Notification
        if (opcode == 0x12) {
          name = "WriteReq";
          relevant = true;
        } else if (opcode == 0x52) {
          name = "WriteCmd";
          relevant = true;
        } else if (opcode == 0x1B) {
          name = "Notify";
          relevant = true;
        }

        if (relevant) {
          // Handle is next 2 bytes
          if (attData.length >= 3) {
            int attHandle = attData[1] | (attData[2] << 8);
            Uint8List val = attData.sublist(3);
            String hex = val
                .map((b) => b.toRadixString(16).padLeft(2, '0').toUpperCase())
                .join("");

            // Filter useless/short packets if needed, or show all
            print(
                "#$packetCount $name [H:0x${attHandle.toRadixString(16)}] : $hex");
          }
        }
      }
    }
  }
}

File: myFilesForDebuggingAndInformations/analyze_btsnoop.py
import struct
import sys
import os

def parse_btsnoop(filepath):
    print(f"--- Analyzing {os.path.basename(filepath)} ---")
    try:
        with open(filepath, "rb") as f:
            header = f.read(16)
            if len(header) < 16:
                print("File too short for header")
                return

            magic = header[:8]
            if magic != b'btsnoop\0':
                print(f"Invalid magic: {magic}")
                # It might be in pcap format instead of btsnoop?
                # Some androids use pcap. Magic for pcap is different.
                # standard pcap: a1 b2 c3 d4 (big endian) or d4 c3 b2 a1 (little)
                # Let's check for "btsnoop"
                return

            version, datalink = struct.unpack(">II", header[8:16])
            print(f"Version: {version}, DataLink: {datalink}")

            packet_count = 0
            while True:
                # Record Header: 24 bytes
                rec_header = f.read(24)
                if len(rec_header) < 24:
                    break

                orig_len, inc_len, flags, drops, time_hi, time_lo = struct.unpack(">IIIIII", rec_header)
                
                # Read Data
                packet_data = f.read(inc_len)
                if len(packet_data) < inc_len:
                    break

                packet_count += 1
                
                # HCI Packet Type (1 byte) is usually the first byte of packet_data for UART (H0) encapsulation
                # Types: 0x01=Cmd, 0x02=ACL Data, 0x03=SCO, 0x04=Event
                # We care about ACL Data (0x02) for ATT payload
                
                # If DataLink is 1002 (HCI UART), first byte is packet type.
                # If DataLink is 1001 (HCI), there is no packet type byte? (Depends on implementation, usually not)
                # Android usually uses H4 (UART) or similar.
                
                hci_type = packet_data[0]
                
                # We want ACL Data (0x02)
                if hci_type == 0x02:
                    # ACL Data Header: Handle (12 bits) + PB (2 bits) + BC (2 bits) | Total Len (2 bytes)
                    # Total 4 bytes usually (excluding type)
                    # packet_data[1] and [2] has Handle info
                    if len(packet_data) < 5: 
                        continue
                        
                    handle_flags = struct.unpack("<H", packet_data[1:3])[0]
                    conn_handle = handle_flags & 0x0FFF
                    
                    # L2CAP Header follows ACL Header
                    # Length (2 bytes) + CID (2 bytes)
                    l2cap_offset = 5 # 1(Type) + 4(ACL Header)
                    if len(packet_data) < l2cap_offset + 4:
                        continue
                        
                    l2cap_len, l2cap_cid = struct.unpack("<HH", packet_data[l2cap_offset:l2cap_offset+4])
                    
                    # Check for ATT (CID = 0x0004)
                    if l2cap_cid == 0x0004:
                        att_offset = l2cap_offset + 4
                        att_data = packet_data[att_offset:]
                        
                        if len(att_data) > 0:
                            opcode = att_data[0]
                            # ATT Opcodes of interest:
                            # 0x12: Write Request
                            # 0x52: Write Command
                            # 0x1B: Handle Value Notification
                            # 0x1D: Handle Value Indication
                            
                            opcode_name = ""
                            is_relevant = False
                            
                            if opcode == 0x12: 
                                opcode_name = "Write Req"
                                is_relevant = True
                            elif opcode == 0x52: 
                                opcode_name = "Write Cmd"
                                is_relevant = True
                            elif opcode == 0x1B: 
                                opcode_name = "Notification"
                                is_relevant = True
                            
                            if is_relevant:
                                # Extract Handle (2 bytes)
                                if len(att_data) >= 3:
                                    att_handle = struct.unpack("<H", att_data[1:3])[0]
                                    value = att_data[3:]
                                    if len(value) > 0:
                                        hex_val = value.hex().upper()
                                        print(f"#{packet_count} {opcode_name} [Handle: 0x{att_handle:04X}]: {hex_val}")

    except Exception as e:
        print(f"Error parsing: {e}")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python analyze_btsnoop.py <file1> [file2...]")
    else:
        for f in sys.argv[1:]:
            if os.path.exists(f):
                parse_btsnoop(f)
            else:
                print(f"File not found: {f}")
